### DISCUSSION TITLE
Berry and LVGL, create/load pages/screens or clearing the display?

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#20570] [Q&A]
I wasn't savvy enough or perhaps too tired. The only example I found that changes pages is the HASPmota LiveDemo. All the Berry examples appear to be "one-page" only, with buttons triggering print commands.

How can I clear the display and create new pages that can be loaded through menus or buttons?
Thanks for all!

**Comment by sfromis** [comment]
While I don't know about how to do specific things with LVGL in Tasmota, what I do know is that the starting point should be based on what you'd do in LVGL outside of Tasmota, and then adapt that to the differences with how LVGL is integrated with Berry in Tasmota. https://tasmota.github.io/docs/LVGL_Internals/

Of potential interest, here's how the Berry code for HASPmota seems to allocate a new screen per page, when having multiple pages:
> `self._lv_scr = lv.obj(0)      # allocate a new screen`

https://github.com/arendst/Tasmota/blob/development/tasmota/berry/haspmota_src/haspmota_core/haspmota.be#L1481

**Comment by s-hadinger** [comment]
What you are looking for is called "Screens" in LVGL: https://docs.lvgl.io/8.3/overview/object.html#screens

HASPmota uses them to create one "Screen" per page and swap them in and out when changing pages.

A screen is indeed create by `lv.obj(0)`, i.e. an object without a parent.

**Comment by ChrisVonFrosta** [comment]
thank you both, i saw this but was not able to adapt it. i gave it some tries but without success.
i started with the exapmple from [here ](https://tasmota.github.io/docs/LVGL_in_10_minutes/#try-a-lvgl-demo-app). The target is to load another page when pressing one of the buttons. Before changing the button behaviar i want to create a new page.

`scr1 = lv.obj_create();`  results in: 
```
 BRY: Exception> 'attribute_error' - module 'lv' has no member 'obj_create'
stack traceback:
	<unknown source>: in function `main`
```

(stupid hacking):
```
print(scr)
lv.scr_load(scr)
<instance: lv_obj(0x3D8CEBF8)>
scr1 = lv.obj_create(0);
BRY: Exception> 'attribute_error' - module 'lv' has no member 'obj_create'
stack traceback:
	<unknown source>: in function `main`
scr1 = lv.obj_create(NIL);
BRY: Exception> 'syntax_error' - input:1: 'NIL' undeclared (first use in this function)
scr1 = lv.obj_create(Nil);
BRY: Exception> 'syntax_error' - input:1: 'Nil' undeclared (first use in this function)
scr1 = lv.obj_create(nil);
BRY: Exception> 'attribute_error' - module 'lv' has no member 'obj_create'
stack traceback:
	<unknown source>: in function `main`
scr1 = lv.obj_create();
BRY: Exception> 'attribute_error' - module 'lv' has no member 'obj_create'
stack traceback:
	<unknown source>: in function `main`
```

the name transformation seems to be not allways just `lv_` to `lv.`.
`lv_screen_load(scr)` seems to be `lv.scr_load(scr)`

bad for me, that the only example with multipage is in HASPmota.
sorry for lot of asking...

**Reply by sfromis** [reply]
What we wrote was `lv.obj(0)` to create a new screen object, not `lv.obj_create(0)`.

**Comment by ChrisVonFrosta** [comment]
really?? i am so sorry that i didn't see that "detail". It works, i will play. thank you! 
But something like that hasn't been documented anywhere yet, has it?

**Reply by s-hadinger** [reply]
There are more than a thousand functions in LVGL, I have no intention to document all them as they are already in LVGL.

However I agree that the documentation could be more explicit.

Now be aware that I am currently porting to LVGL 9 so some details may change.

**Reply by s-hadinger** [reply]
Btw, the message `module 'lv' has no member 'obj_create'` is explicit enough, you don't need to try different parameters, the function just does not exist.

**Reply by sfromis** [reply]
What can be confusing in this case is that LVGL docs has `lv_obj_create` which by the "usual" name transformation scheme would become `lv.obj_create` instead of `lv.obj`.

**Reply by s-hadinger** [reply]
Yes, that's the exception (it's the same with MicroPython). Since Berry is object oriented, there is a proper constructor and no create function.

Actually our documentation is outdated, it's now `lv.obj(0)` and not `lv_obj(0)` as stated in the doc.

**Reply by sfromis** [reply]
I was wondering whether it would be feasible to transform the binding mapping specifications to a "nice" human-readable form. It does look reasonably systematic, but not knowing "anything" about LVGL, I can't say that I understand the details of how the mapping works.
https://github.com/arendst/Tasmota/blob/development/lib/libesp32_lvgl/lv_binding_berry/generate/be_lv_c_mapping.h

**Reply by ChrisVonFrosta** [reply]
> There are more than a thousand functions in LVGL, I have no intention to document all them as they are already in LVGL.
> 
> However I agree that the documentation could be more explicit.
> 
> Now be aware that I am currently porting to LVGL 9 so some details may change.

That's of course understandable. Interesting about LVGL9. Will you also be implementing its name changes?
`scr` --> `screen` , `lv.btn` --> `lv.button`

**Reply by s-hadinger** [reply]
Yes but I'm implementing some synonyms so most of the LVGL8 names will continue to work with LVGL9

**Reply by s-hadinger** [reply]
@sfromis It would be indeed possible to generate automatic list of objects and methods with input/output parameters - but sadly getting the detail of what each method does is not simple. I could just quote the native LVGL method

**Reply by sfromis** [reply]
What I meant was not documentation with descriptions, but a listing to make the mapping clear, LVGL name xx becomes yy in Tasmota. While the `lv_` to `lv.` rule covers a lot, it gets more tricky when not knowing about exceptions like `lv_obj_create`, or even differences due to LVGL versioning.

Sure, function description could be added, but that would indeed just be quoting the LVGL docs.

Where I'm coming from is liking to have a path to find things, without having to already know about both sides. I've also found that extracting such lists can help spot places where things are not as consistent as they should.

**Reply by s-hadinger** [reply]
Ok, it should be quite easy to generate markdown for all the existing mappings. Let's try if it's any useful

**Reply by s-hadinger** [reply]
I did a proof-of-concept for such list of functions. It's not complete nor are the arguments explicit, but it gives an overview of what it could look like: https://tasmota.github.io/docs/LVGL_API_Reference/

**Reply by sfromis** [reply]
Looks good. It being searchable adds value to having it in the docs. And it also reveals oddities like having alias names which otherwise could be confusing to find:
> `set_angle	lv_scale_set_rotation`

And yes, I also saw the "straight" binding
> `set_rotation	lv_scale_set_rotation`

**Reply by s-hadinger** [reply]
Sorry I forgot to mention this is the documentation for LGVL 9. `set_angle` is renamed to `set_rotation` but I added synonyms to avoid breaking changes

**Reply by s-hadinger** [reply]
Unfortunately it's not possible to generate a link to the entry in the LVGL documentation, links are not predictible

**Reply by sfromis** [reply]
Yeah, I noticed a lot of those aliased synonyms, 74 LVGL names matched by 2 or even 3 Tasmota names. One oddity I saw was these bindings - LVGL name followed by two Tasmota names, the double underscore due to the concatenation (to have matchable names) with the md table containing names starting with underscore:
```
lv_buttonmatrix_get_button_text lv_buttonmatrix__btn_text lv_buttonmatrix_get_button_text
lv_keyboard_get_button_text     lv_keyboard__btn_text     lv_keyboard_get_button_text
lv_list_get_button_text         lv_list__btn_text         lv_list_get_button_text
```

When it comes to linking to LVGL docs, one approach could be to generate generic search links. Even with the inherent fuzzyness of such searches, it does appear to reasonably useful, like:
https://docs.lvgl.io/master/search.html?q=lv_scale_set_rotation&check_keywords=yes&area=default
